kotlin/src/test/java/com/squareup/moshi/KotlinJsonAdapterTest.kt

```diff
@@ -296,6 +296,39 @@ class KotlinJsonAdapterTest {
     var b: Int = -1
   }
 
+  @Test fun doesNotAssignToPropertiesOfSameNameAsConstructorParameters() {
+    val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
+    val jsonAdapter = moshi.adapter(ConstructorParameterWithSameNameAsProperty::class.java)
+
+    val encoded = ConstructorParameterWithSameNameAsProperty(5)
+    assertThat(jsonAdapter.toJson(encoded)).isEqualTo("{\"a\":-1,\"b\":5}")
+
+    val decoded = jsonAdapter.fromJson("{\"a\":5}")!!
+    assertThat(decoded.a).isEqualTo(-1)
+    assertThat(decoded.b).isEqualTo(5)
+  }
+
+  class ConstructorParameterWithSameNameAsProperty(a: Int) {
+    var a = -1
```

Comment:
I have a few concerns about this behavior that I want to check. Like what happens with type qualifiers or transient?
I'll look into it, ask questions, and follow up.